apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/TransactionTests.cs

```diff
@@ -475,6 +476,33 @@ public async Task ReadStrongSingle()
         }
         // [END spanner_test_strong_read_concurrent_updates]
 
+        // [START spanner_test_read_exact_with_return_read_timestamp]
+        [Fact]
+        public async Task ReadExactWithReturnReadTimestamp()
+        {
+            using (var connection = _fixture.GetConnection())
+            {
+                await connection.OpenAsync();
+                var targetReadTimestamp = _fixture.TimestampBeforeEntries;
+                var bound = TimestampBound.OfReadTimestamp(targetReadTimestamp);
+                bound.ReturnReadTimestamp = true;
+                using (var tx =
+                    await connection.BeginReadOnlyTransactionAsync(bound))
+                {
+                    Assert.Equal(TransactionMode.ReadOnly, tx.Mode);
+                    Assert.Equal(targetReadTimestamp, tx.TimestampBound.Timestamp);
+
+                    var cmd = CreateSelectAllCommandForKey(connection);
+                    cmd.Transaction = tx;
+                    using (var reader = await cmd.ExecuteReaderAsync())
+                    {
+                        Assert.Equal(reader.GetReadTimestamp(), wkt::Timestamp.FromDateTime(targetReadTimestamp));
```

Comment:
Note that this line throws a null reference error because `PopulateMetadata().Transaction` doesn't exist. It might be because there's no data being returned by this reader or for some other reason. I need to debug.

Reply:
Happy to help debug if you'd like me to. Just give me a shout.

Reply:
You need to call `reader.Read()` or `reader.ReadAsync()` at least once before any (meta)data is returned.

Reply:
Ah that old problem! Nice catch, thanks @olavloite :)

Turns out the timestamp is also unavailable if your read doesn't return any results so I had to change the test a bit.

Reply:
Eek - is that "also unavailable" part of this implementation somewhere lower level, or is it general Spanner-wide? It's a surprise to me.

Reply:
> Turns out the timestamp is also unavailable if your read doesn't return any results so I had to change the test a bit.

No, that should not be the case. Cloud Spanner always returns a read timestamp if your read-only transaction requests one, regardless whether the query returns any rows or not. So I think that this is a problem in the C# client in that case. If you execute a query on Cloud Spanner that requests a timestamp and the query does not return any results, Cloud Spanner will always return a `PartialResultSet` containing the metadata of the query. That metadata will include the columns in the query (also when there are no results), and the transaction metadata including the read timestamp.
I think the problem might be that the metadata are not read at all by the client if there are no rows in the result set.

(Just tested the above with the Java client to be absolutely sure, and I can verify that Cloud Spanner returns the read timestamp also when the query returns 0 rows)

Reply:
That's encouraging - I'm far happier for this to be a problem in the C# code, as we can fix that :) @skuruppu let me know if you'd like me to take a look diagnostically; otherwise, I'm happy to let you figure it out :)

Reply:
@olavloite thanks for checking this. I'll do a dig around the code to see why we're not getting the metadata.

Reply:
I realize this had nothing to do with whether data is returned. It's because of the type of transaction.

If you try to do an exact read (with a BeginReadOnlyTransactionAsync call as [here](https://github.com/googleapis/google-cloud-dotnet/blob/13ed74b0e70f5b9932c9f679a3378493586c2fcc/apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/TransactionTests.cs#L302), then you can't pass the bound to `ExecuteReaderAsync()` ([link](https://github.com/googleapis/google-cloud-dotnet/blob/13ed74b0e70f5b9932c9f679a3378493586c2fcc/apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/TransactionTests.cs#L308)). Without the bound, the transaction doesn't know that you want to return the timestamp. If you do try to pass it into `ExecuteReaderAsync()`, the client returns https://github.com/googleapis/google-cloud-dotnet/blob/63f54b5ebb8c2f7eec26f0fd7fabfc77e1237596/apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommand.ExecutableCommand.cs#L122

I don't know if this is a case where a user might expect to be able to return the timestamp.

Reply:
(Maybe it's already working, but I noticed that there's no sample or test case for it in the current PR)

There are two ways that a user should be able to request and get a read timestamp:
1. Using a single-use read-only transaction. This is done by passing in a `TimestampBound` to the `SpannerCommand.ExecuteReaderAsync(..)` method.
2. Using a multi-use read-only transaction. This is done by passing in a `TimestampBound` to the `SpannerConnection.BeginReadOnlyTransaction(..)`. The first read operation on the transaction (i.e. the first call to `SpannerCommand.ExecuteReaderAsync()` with a `SpannerCommand` that uses the transaction) should return the read timestamp.

So it should not be necessary to pass in any `TimestampBound` to the `SpannerCommand.ExecuteReaderAsync()` when you are using a multi-use read-only transaction.

Reply:
Ah thanks for the context. I added back the test that wasn't working, which is the multi-use case. I also added a debug print so when we get the test failure, we can see in the logs what is returned in the metadata. You can see it in the emulator test failure in the PR. I haven't yet tried to debug why this is.

Reply:
Sorry, what I meant was that with multi-use read-only transactions, the read timestamp is __not__ returned as part of the metadata of the ResultSet. Instead you have to get it from the transaction that is returned when `BeginTransaction` is called. That is hidden away somewhat in the session pool, which makes it a little harder to get to.
I took the liberty of adding a commit to this PR that shows how you could register and get it in the `DataReader`, but feel free to change the design of it.

(I also have a couple of worries regarding the current implementation of `GetReadTimestamp` on `DataReader`, but I'll add separate comments for those)

Reply:
Thanks @olavloite. I think your design works well. The only thing I changed was for `PooledSession.WithTransaction()` to take a null read timestamp by default as it didn't make much sense for RW transactions to accept a timestamp. But let me know if I misunderstood something there.